### PR TITLE
[WIP] MBE Bug Fixes

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR fixed following bugs :

* Fixed a bug which make expansion fail (By changing `Vec<Bindings>` to `FxHashMap`): 
```rust
macro_rules foo! {
  ($( ($($a:ident)* => $($b:ident))*))* => { ... }
}
foo!( ( => a) (b => ) )
```
* Allow `macro_rules` in `macro_rules` by introduce `MacroRuleChecker`
* Add macro_call pattern parsing in `ra_syntax`.
* Skip dollars in `Parser::bump_raw`
* Add spacing when convert `tt` to `SyntaxNode`
* Make `vis` matcher optional
* Fixed `assert_expansion` to allow compare syntax tree without spacing.
* Move all mbe tests to separate file.

So all dogfooding bugs in mbe are fixed except `DocComment` to $meta matcher, which will be handled in another PR.

### CODE REVIEW

**Comment by edwin0cheng** [comment]
Add more comments and remove unused code.

**Comment by matklad** [comment]
@edwin0cheng is it possible to split this PR in chunks? 

Diff is a bit too large to review effectively :)

File: crates/ra_mbe/src/tests.rs [hunk 1]
```diff
@@ -0,0 +1,1364 @@
+#![cfg(test)]
```
Comment (matklad) [inline crates/ra_mbe/src/tests.rs]: We usually place cfg on `mod test`

**Comment by edwin0cheng** [comment]
>@edwin0cheng is it possible to split this PR in chunks?
> Diff is a bit too large to review effectively :)
Sure, let me drop this PR and make it one per one :)

File: crates/ra_mbe/src/mbe_expander.rs [hunk 2]
```diff
@@ -80,7 +85,7 @@ struct Bindings {
 #[derive(Debug)]
 enum Binding {
     Simple(tt::TokenTree),
-    Nested(Vec<Binding>),
+    Nested(FxHashMap<usize, Binding>),
```
Comment (matklad) [inline crates/ra_mbe/src/mbe_expander.rs]: IIRC, `rustc` is using a `Vec` here: 

https://github.com/rust-lang/rust/blob/f843ad60ef5a78d8b4da85c5007d3ecac229f1a8/src/libsyntax/ext/tt/macro_parser.rs#L339-L359

I think we should probably follow what rustc is doing there .

Comment (edwin0cheng) [reply crates/ra_mbe/src/mbe_expander.rs]: The reason i changed that is, when expanding, there is possible the index is not missing : 
Imagine this case :
```
macro_rules foo! {
  ($( ($($a:ident)* => $($b:ident))*))* => { ... }
}
foo!( ( => a) (b => ) )
```
Let me read how rustc handle that case first, but i suspect that using another method i don't know :)

Comment (edwin0cheng) [reply crates/ra_mbe/src/mbe_expander.rs]: Okay, i got it now, `rustc` implementation push a empty bindings in it if there are zero match repeat. I will follow that.

**Pull request closed without merging**